experiment/ootemplate/http.go

```diff
@@ -0,0 +1,231 @@
+		ootrip.Failure = err.Error()
+		return resp, err
+	}
+	ootrip.Response.Code = resp.StatusCode
+	ootrip.Response.Headers = getheaders(resp.Header)
+	err = getbody(&resp.Body, &ootrip.Response.Body)
+	if err != nil {
+		ootrip.Failure = err.Error()
+		return resp, err
+	}
+	return resp, nil
+}
+
+// NewHTTPClientWithMeasurer is like httpx.NewTracingProxying client except
+// that it also returns a measurer where you can see round trips.
+//
+// The |log| argument is an apex/log.Log compatible logger. The |proxy|
+// argument is the function that should be used to setup a proxy (we don't
+// need a proxy when measuring, except for circumvention tools). The
+// |tlsConfig| argument is an optional TLS configuration with which you
+// can, e.g., supply the path to an extrnal CA bundle.
+func NewHTTPClientWithMeasurer(
+	logger log.Logger, proxy func(req *http.Request) (*url.URL, error),
+	tlsConfig *tls.Config,
+) (client *http.Client, measurer *HTTPMeasurer) {
+	// This creates a transport that contains as transport the measurer in
+	// httptracex, which routes events to the |Handler| and uses as a
+	// real transport |RoundTripper|. Here |RoundTripper| is a real HTTP
+	// transport. The httptracex.Measurer is currently only used for
+	// logging via httplog.RoundTripLogger. We have improvements in
+	// ooni/probe-engine#26 that will allow us to collect much more low level
+	// data than now (e.g. TLS). For now the plan is to just tap into the
+	// round-trip to collect the bare minimum required to implement Telegram.
+	//
+	// TODO(bassosimone): modify the code such that we can take advantage
+	// of the httptracex framework to log everything HTTP.
+	measurer = &HTTPMeasurer{
+		RoundTripper: &httptracex.Measurer{
+			RoundTripper: httpx.NewTransport(proxy, tlsConfig),
+			Handler: &httplog.RoundTripLogger{
+				Logger: logger,
+			},
+		},
+	}
+	// This creates an HTTP client where the transport is our measurer
+	client = &http.Client{Transport: measurer}
+	return
+}
+
+// HTTPRequesTemplate is a quick template for setting up an HTTP request
+// for performing OONI measurements.
+type HTTPRequestTemplate struct {
+	Method    string
+	URL       string
+	UserAgent string
+}
+
+// HTTPPerformMany performs the many HTTP requests described by the
+// provided template using a custom client and then returns the measurements
+// as a list of HTTP round trips. This function returns an error when we
+// cannot even prepare a request because you provided us with invalid data,
+// otherwise in any other case it will succeed.
+func HTTPPerformMany(
+	ctx context.Context, logger log.Logger, templates ...HTTPRequestTemplate,
+) ([]HTTPRoundTrip, error) {
+	var requests []*http.Request
+	for _, t := range templates {
+		req, err := http.NewRequest(t.Method, t.URL, nil)
+		if err != nil {
+			return nil, err
+		}
+		if t.UserAgent == "" {
+			// 11.8% as of August 24, 2019 according to
+			// https://techblog.willshouse.com/2012/01/03/most-common-user-agents/
```

Comment:
@hellais we probably want to distribute user agents using orchestra, don't we?

Reply:
Possibly, though we ought to still have a sane default in case the connection to the orchestra fails.

Reply:
Yes, sure, I'm chiefly worried that we'll not update the header in a long time.